doc/IdleHandler.3
'\"
'\" Copyright (c) 1998 Sun Microsystems, Inc.
'\"
'\" See the file "license.terms" for information on usage and redistribution
'\" of this file, and for a DISCLAIMER OF ALL WARRANTIES.
'\" 
'\" RCS: @(#) $Id$
'\" 
.so man.macros
.TH IdleHandler 3 1.0 Jacl "Jacl Library Procedures"
.BS
.SH CLASS
tcl.lang.IdleHandler -- Subclass from this abstract base class to
create idle event handlers.

.SH METHODS
.sp
\fBIdleHandler\fR(Notifier \fInotifier\fR)
.sp
synchronized void \fBcancel\fR()	
.sp
abstract void \fBprocessIdleEvent\fR()

.SH ARGUMENTS
.AS TclObject "table[]" .
.AP Notifier notifier
The Notifier in which to register the new IdleHandler.
.BE

.SH DESCRIPTION

The \fBIdleHandler\fR class is used to schedule code to be executed
when the specified \fInotifier\fR becomes idle. The \fInotifier\fR is
considered to be idle when \fInotifier\fR.\fBdoOneEvent\fR has been
called, couldn't find any events to handle, and is about to go to
sleep waiting for an event to occur. At this point all pending idle
handlers are invoked. For each currently registered IdleHandler
instance, there will be a single call to its \fBprocessIdleEvent\fR
method; after that the IdleHandler instance is automatically
removed. The IdleHandler class is only usable in programs that use
\fInotifier\fR.\fBdoOneEvent\fR to dispatch events.

.TP
\fBIdleHandler\fR
This constructor creates an IdleHandler instance and register it
inside the given \fInotifier\fR.
.TP
\fBcancel\fR
If the \fBprocessIdleEvent\fR method has not been invoked yet, cancels
this idle handler. If the \fBprocessIdleEvent\fR method has already
been invoked, this call has no effect.
.TP
\fBprocessIdleEvent\fR
This method is called when the \fInotifier\fR is idle. You can
override this method to perform the desired actions when the idle
event occurs. The IdleHandler instance can be created in any
thread. However, \fBprocessIdleEvent\fR is always called in the
primary thread of the \fInotifier\fR that was specified in the
IdleHandler constructor.

.SH "EQUIVALENT C FUNCTIONS"
Tcl_CreateIdleHandler

.SH "SEE ALSO"
Notifier, TimerHandler

.SH KEYWORDS
idle, event, event queue


ChangeLog
2019-05-04  Fabrice Ducos <fabrice.ducos@gmail.com>

        * Support of latest versions of Tcl (at least Tcl 8.5
        and Tcl 8.6) and of modern JDK (10+);
        support of JDK 4- abandoned (use of auto-boxing);
        support of OSX systems added;
        details below

        * configuration (tcljava.m4, configure.in),
        and compilation (Makefile.in) templates updated for
        JDK 10+ (javah not available anymore).
        javah is still used when it is available (JDK 9-).
        * minor update of jtclsh.in (for OSX dynamic libraries)
        * many obsolescent constructs (mostly old-style boxing)
        removed in *.java under src/jacl/tcl/lang,
        src/tclblend/tcl/lang and src/tcljava/tcl/lang,
        in order to silence many warnings.
        (effect: support of JDK 4- abandoned)
        * src/native/javaInterp.c: 2010 implementation
        of Java_tcl_lang_Interp_initName() crashed on the
        example given at docs/website/getstart.html,
        because calling Tcl_GetNameOfExecutable
        before Tcl_FindExecutable gives rise to undefined
        behaviour, at least in latest versions of Tcl
        (crashes observed on several systems, at least with
        Tcl 8.6).
        Java_tcl_lang_Interp_initName() is now simplified and
        always calls Tcl_FindExecutable first.
        * all the occurences of interp->result have been
        replaced with Tcl_GetStringResult(interp).
        Accessing interp->result directly was deprecated in
        latest versions of the Tcl C API, and removed from
        Tcl 8.6 (another solution was to #define USE_INTERP_RESULT,
        but this would have been an easy-going, quick and dirty way)
        ref. https://www.tcl.tk/man/tcl/TclLib/Interp.htm

2019-05-01  Fabrice Ducos <fabrice.ducos@gmail.com>

        * tcljava resurrected
        * migration from a CVS to a GIT repository
        (with reposurgeon)

2010-02-26  Mo DeJong  <mdejong@users.sourceforge.net>

	* tests/tcl/regexp.test: Add test case for
        escaped - character in regexp command.
        [Bug 2012389]

2010-02-21  Mo DeJong  <mdejong@users.sourceforge.net>

	* src/jacl/tcl/lang/Regex.java (replaceAll):
	Fixup regsub -all logic so 5 more tests pass.
	The replaceAll() logic is broken because it depends
	on matching substrings instead of using an index
	into the larger string and Java's regexp library
	uses this to calculate line starts in multiline mode.

2010-02-18  Mo DeJong  <mdejong@users.sourceforge.net>

        Incorporate fix/tests from Tcl head on 2010-02-11,
        aka was Tcl [Bug 2826551, Patch 2948425].

	* src/jacl/tcl/lang/RegexpCmd.java (cmdProc):
        Fixup increment index on zero length match issue.
	* tests/tcl/regexp.test: Back port of new regexp tests,
        some tests marked with {endIndex} due to lack of impl.

2010-02-11  Mo DeJong  <mdejong@users.sourceforge.net>

	Patch 2946118

	* src/jacl/tcl/lang/ListCmd.java: Minor style reformat.
	* src/jacl/tcl/lang/Util.java (scanElement, convertElement):
	Remove crusty old code that treated backslash zero
	as the end of the line. That logic was removed from
	C Tcl long ago.
	* tests/inprogress/EventAdaptor.test: Fixup tests
	that depended on backslash zero quirks.
	* tests/tcl/list.test: Add test cases for backslash zero
	as argument to list command.

2009-10-04  Mo DeJong  <mdejong@users.sourceforge.net>

	* src/jacl/tcl/lang/Regex.java (match, _match,
        replaceFirst, replaceAll): Rewrite -start offset
        logic to correctly deal with odd edge cases.
	* tests/tcl/regexp.test: Add additional test cases
        that cover -start index edge cases.

2009-10-04  Mo DeJong  <mdejong@users.sourceforge.net>

	* src/jacl/tcl/lang/RegexpCmd.java (cmdProc):
	* src/jacl/tcl/lang/RegsubCmd.java (cmdProc):
        Rewrite argument processing logic to avoid using
        IndexOutOfBoundsException to deal with an incorrect
        number of arguments.

2009-10-03  Mo DeJong  <mdejong@users.sourceforge.net>

	* src/jacl/tcl/lang/Regex.java (replaceFirst, replaceAll):
        Fix regsub -all crash because offset was not getting
        advanced by 1 on a match of length zero.
        Fixup -start argument support with and without -all
        in the case where the offset was larger than the
        length of the string.
	* tests/tcl/regexp.test: Add additional test cases.
        Enable regsub test cases that were getting caught
        in an infinite loop.

2009-09-30  Mo DeJong  <mdejong@users.sourceforge.net>

	* src/jacl/tcl/lang/RegexpCmd.java (cmdProc): Fixup
        failing tests related to optional subgroups.
	* src/jacl/tcl/lang/RegsubCmd.java (cmdProc):
        Minor paren fix, no function changes.

2009-09-30  Mo DeJong  <mdejong@users.sourceforge.net>

	* src/jacl/tcl/lang/RegexpCmd.java (cmdProc): Rework
        logic related to match result groups.
	* tests/tcl/regexp.test: Add missing test cases for
        non-capturing groups and -inline results.

2009-09-26  Mo DeJong  <mdejong@users.sourceforge.net>

	* src/jacl/tcl/lang/Regex.java (replaceFirst, replaceAll):
        Fixup use of MULTILINE flag.
	* src/jacl/tcl/lang/RegexpCmd.java (cmdProc): Fixup
        support for -linestop and -lineanchor arguments to regexp.
	* src/jacl/tcl/lang/RegsubCmd.java (cmdProc):  Fixup
        support for -linestop and -lineanchor arguments to regsub.
	* tests/tcl/regexp.test: Add missing test cases for
        -linestop and -lineanchor arguments.

2009-09-22  Mo DeJong  <mdejong@users.sourceforge.net>

	* src/jacl/tcl/lang/Regex.java (replaceFirst, replaceAll):
        Fixup support for the empty string as the string
        argument to regsub by removing the MULTILINE flag when
        an empty string is detected.
	* src/jacl/tcl/lang/RegexpCmd.java (cmdProc): Fixup
        support for empty string as string argument.
	* src/jacl/tcl/lang/RegsubCmd.java (cmdProc):
        Remove comments.
	* tests/tcl/regexp.test: Add additional test cases

tcljava/src/tclblend/tcl/lang/IdleHandler.java
/* 
 * IdleHandler.java --
 *
 *	This class is used to create handlers for idle events.
 *
 * Copyright (c) 1998 by Sun Microsystems, Inc.
 *
 * See the file "license.terms" for information on usage and redistribution
 * of this file, and for a DISCLAIMER OF ALL WARRANTIES.
 *
 * RCS: @(#) $Id: IdleHandler.java,v 1.1.1.1 1998/10/14 21:09:10 cvsadmin Exp $
 */

package tcl.lang;

abstract public class IdleHandler {

// True if the cancel() method has been called.

boolean isCancelled;

// Global ref passed to Tcl, read/written from C

long clientData;


/*
 *----------------------------------------------------------------------
 *
 * IdleHandler --
 *
 *	Create a idle handler to be fired when the notifier is idle.
 *
 * Results:
 *	None.
 *
 * Side effects:
 *	The idle is registered in the list of idle handlers in the
 *	given notifier. When the notifier is idle, the
 *	processIdleEvent() method will be invoked exactly once inside
 *	the primary thread of the notifier.
 *
 *----------------------------------------------------------------------
 */

public
IdleHandler(
    Notifier n)			// The notifier to fire the event.
{
    isCancelled = false;
    clientData = doWhenIdle();
}

/*
 *----------------------------------------------------------------------
 *
 * cancel --
 *
 *	Mark this idle handler as cancelled so that it won't be invoked.
 *
 * Results:
 *	None.
 *
 * Side effects:
 *	The idle handler is marked as cancelled so that its
 *	processIdleEvent() method will not be called. If the idle
 *	event has already fired, then nothing this call has no effect.
 *
 *----------------------------------------------------------------------
 */

public synchronized void
cancel()
{
    if (isCancelled) {
	return;
    }

    isCancelled = true;
    cancelIdleCall(clientData);
    clientData = 0;
}

/*
 *----------------------------------------------------------------------
 *
 * invoke --
 *
 *	Execute the idle handler. This method is called by the
 *	native method JavaIdleProc only.
 *
 *	Because the idle handler may be being cancelled by another
 *	thread, both this method and cancel() must be synchronized to
 *	ensure correctness.
 *
 * Results:
 *	None.
 *
 * Side effects:
 *	The idle handler may have arbitrary side effects.
 *
 *----------------------------------------------------------------------
 */

synchronized final void
invoke()
{
    if (isCancelled) {
	throw new TclRuntimeError("IdleHandler.invoke(): event was cancelled");
    }
    processIdleEvent();
    clientData = 0;
}

/*
 *----------------------------------------------------------------------
 *
 * processIdleEvent --
 *
 *	This method is called when the idle is expired. Override
 *	This method to implement your own idle handlers.
 *
 * Results:
 *	None.
 *
 * Side effects:
 *	It can do anything.
 *
 *----------------------------------------------------------------------
 */

abstract public void
processIdleEvent();

/*
 *----------------------------------------------------------------------
 *
 * doWhenIdle --
 *
 *	Create a C level idle handler for this object.
 *
 * Results:
 *	Returns a pointer to store in clientData.
 *
 * Side effects:
 *	None.
 *
 *----------------------------------------------------------------------
 */

private final native long
doWhenIdle();

/*
 *----------------------------------------------------------------------
 *
 * cancelIdleCall --
 *
 *	Delete a C level idle handler.
 *
 * Results:
 *	None.
 *
 * Side effects:
 *	None.
 *
 *----------------------------------------------------------------------
 */

private final native void
cancelIdleCall(long clientData);

} // end IdleHandler

